src/Hl7.Fhir.Base/FhirPath/Expressions/EchoVisitor.cs

```diff
@@ -0,0 +1,301 @@
+        }
+
+        private void OutputPrecedingTokens(Expression? expr)
+        {
+            if (expr?.leadingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", expr.leadingWhitespace.Select(ws => ws.ToString())));
+        }
+        private void OutputTrailingTokens(Expression expr)
+        {
+            if (expr.trailingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", expr.trailingWhitespace.Select(ws => ws.ToString())));
+        }
+
+        private void OutputSubToken(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            OutputPrecedingTokens(subtoken);
+            _result.Append($"{subtoken.Value}");
+            OutputTrailingTokens(subtoken);
+        }
+
+        private void OutputPrecedingTokens(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            if (subtoken.leadingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", subtoken.leadingWhitespace.Select(ws => ws.ToString())));
+        }
+        private void OutputTrailingTokens(SubToken subtoken)
+        {
+            if (subtoken == null) return;
+            if (subtoken.trailingWhitespace?.Any() == true)
+                _result.Append(System.String.Join("", subtoken.trailingWhitespace.Select(ws => ws.ToString())));
+        }
+        #endregion
+
+        public override StringBuilder VisitConstant(ConstantExpression expression)
+        {
+            OutputPrecedingTokens(expression);
+            if (expression is IdentifierExpression identifier)
+            {
+                OutputIdentifierName(identifier.Value);
+                OutputTrailingTokens(expression);
+                return _result;
+            }
+            var t = Fhir.Serialization.PrimitiveTypeConverter.ConvertTo<string>(expression.Value);
+
+            switch (expression.ExpressionType.Name)
+            {
+                case "Any":
+                    _result.Append($"{t}");
+                    break;
+                case "Boolean":
+                    _result.Append($"{t}");
+                    break;
+                case "Code":
+                    _result.Append($"{t}");
+                    break;
+                case "Concept":
+                    _result.Append($"{t}");
+                    break;
+                case "Date":
+                    _result.Append($"@{t}");
+                    break;
+                case "DateTime":
+                    if (t.Contains('T'))
+                        _result.Append($"@{t}");
+                    else
+                        _result.Append($"@{t}T");
+                    break;
+                case "Decimal":
+                    _result.Append($"{t}");
+                    break;
+                case "Integer":
+                    _result.Append($"{t}");
+                    break;
+                case "Long":
```

Comment:
You could combine all cases that do the same thing, to make clear they are doing the same thing!